daemon/oci_linux.go

```diff
@@ -751,37 +768,56 @@ func (daemon *Daemon) createSpec(c *container.Container) (retSpec *specs.Spec, e
 	}
 
 	if err := daemon.initCgroupsPath(parentPath); err != nil {
-		return nil, fmt.Errorf("linux init cgroups path: %v", err)
+		// all of the errors in initCgroupsPath are things like failing to make
+		// a directory, or are otherwise the result of a function call with
+		// hard-coded arguments, so return a System error if it fails.
+		return nil, errdefs.System(fmt.Errorf("linux init cgroups path: %v", err))
 	}
 	if err := setDevices(&s, c); err != nil {
-		return nil, fmt.Errorf("linux runtime spec devices: %v", err)
+		// The errors resulting from devices are generally more of a
+		// user-configured-it-wrong error, so return InvalidParameter
+		return nil, errdefs.InvalidParameter(fmt.Errorf("linux runtime spec devices: %v", err))
 	}
 	if err := daemon.setRlimits(&s, c); err != nil {
-		return nil, fmt.Errorf("linux runtime spec rlimits: %v", err)
+		// as of this writing, setRlimits doesn't return any errors at all, so
+		// we return error type Unknown if it happens.
+		return nil, errdefs.Unknown(fmt.Errorf("linux runtime spec rlimits: %v", err))
 	}
 	if err := setUser(&s, c); err != nil {
-		return nil, fmt.Errorf("linux spec user: %v", err)
+		// setUser fails if you specify a bad user, so it's an InvalidParameter
+		return nil, errdefs.InvalidParameter(fmt.Errorf("linux spec user: %v", err))
 	}
 	if err := setNamespaces(daemon, &s, c); err != nil {
-		return nil, fmt.Errorf("linux spec namespaces: %v", err)
+		// setNamespaces has been fixed to return all errdefs errors, so we
+		// don't need to choose one here. additionally, the errdefs package is
+		// aware of the "causer" interface from github.com/pkg/errors, so we
+		// can safely wrap this error so that the actual error message still
+		// starts with "linux spec namespaces"
+		return nil, errors.Wrap(err, "linux spec namespaces")
 	}
 	capabilities, err := caps.TweakCapabilities(oci.DefaultCapabilities(), c.HostConfig.CapAdd, c.HostConfig.CapDrop, c.HostConfig.Capabilities, c.HostConfig.Privileged)
 	if err != nil {
-		return nil, fmt.Errorf("linux spec capabilities: %v", err)
+		// caps.TweakCapabilities also uses all errdefs errors
+		return nil, errors.Wrap(err, "linux spec capabilities")
 	}
 	if err := oci.SetCapabilities(&s, capabilities); err != nil {
-		return nil, fmt.Errorf("linux spec capabilities: %v", err)
+		// SetCapabilities returned no error in this revision, so if it fails,
+		// we don't know why
+		return nil, errdefs.Unknown(fmt.Errorf("linux spec capabilities: %v", err))
```

Comment:
I see there's some more `fmt.Errorf(`'s here (and below); should all ideally be changed to `errors.Wrap` so that the original error doesn't get lost.